File: app/src/main/java/com/tkdev/weatherapp/common/WeatherRetrofitImpl.java
package com.tkdev.weatherapp.common;

import com.tkdev.weatherapp.current.model.WeatherRetrofit;
import com.tkdev.weatherapp.forecast.model.ForecastRetrofit;


import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.tkdev.weatherapp.common.RetrofitCalls.WEATHER_REQUEST_BASE;

public class WeatherRetrofitImpl implements com.tkdev.weatherapp.common.MainContract.Model {

    com.tkdev.weatherapp.common.RetrofitService service;


    @Override
    public void getWeather(com.tkdev.weatherapp.common.MainContract.APIListener callback, String city) {

        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(WEATHER_REQUEST_BASE)
                .addConverterFactory(GsonConverterFactory.create())
                .client(okHttpClient)
                .build();

        service = retrofit.create(com.tkdev.weatherapp.common.RetrofitService.class);

        Call<WeatherRetrofit> call = service.getCurrentWeather(city);

        call.enqueue(new Callback<WeatherRetrofit>() {
            @Override
            public void onResponse(Call<WeatherRetrofit> call, Response<WeatherRetrofit> response) {

                if (response.body() != null && response.isSuccessful()) {


                    callback.onSuccessResponse(response);

                } else {
                    callback.onFailureResponse(response.message());
                }
            }

            @Override
            public void onFailure(Call<WeatherRetrofit> call, Throwable t) {
            }

        });
    }


    @Override
    public void getWeather(com.tkdev.weatherapp.common.MainContract.APIForecastListener callback, String city) {
        HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(WEATHER_REQUEST_BASE)
                .addConverterFactory(GsonConverterFactory.create())
                .client(okHttpClient)
                .build();

        service = retrofit.create(com.tkdev.weatherapp.common.RetrofitService.class);

        Call<ForecastRetrofit> call = service.getForecastWeather(city);

        call.enqueue(new Callback<ForecastRetrofit>() {
            @Override
            public void onResponse(Call<ForecastRetrofit> call, Response<ForecastRetrofit> response) {
                if (response.body() != null && response.isSuccessful()) {

                    callback.onSuccessResponse(response);

                } else {
                    callback.onFailureResponse(response.message());
                }

            }

            @Override
            public void onFailure(Call<ForecastRetrofit> call, Throwable t) {

            }
        });


    }


}

File: app/src/main/java/com/tkdev/weatherapp/common/WeatherRetrofitImpl.kt
package com.tkdev.weatherapp.common

import com.tkdev.weatherapp.current.model.WeatherRetrofit
import com.tkdev.weatherapp.forecast.model.ForecastRetrofit
import com.tkdev.weatherapp.common.MainContract.APIForecastListener
import com.tkdev.weatherapp.common.MainContract.APIListener
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class WeatherRetrofitImpl : MainContract.Model {
    private var service: RetrofitService

    init {
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
        val okHttpClient = OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .build()
        val retrofit = Retrofit.Builder()
                .baseUrl(RetrofitCalls.WEATHER_REQUEST_BASE)
                .addConverterFactory(GsonConverterFactory.create())
                .client(okHttpClient)
                .build()
        service = retrofit.create(RetrofitService::class.java)

    }

    override fun getWeather(listener: APIListener, city: String) {

        val call = service.getCurrentWeather(city)
        call.enqueue(object : Callback<WeatherRetrofit> {
            override fun onResponse(call: Call<WeatherRetrofit>, response: Response<WeatherRetrofit>) {
                if (response.body() != null && response.isSuccessful) {
                    listener.onSuccessResponse(response)
                } else {
                    listener.onFailureResponse(response.message())
                }
            }

            override fun onFailure(call: Call<WeatherRetrofit>, t: Throwable) {
                t.message?.let { listener.onFailureResponse(it) }
            }
        })
    }

    override fun getWeather(listener: APIForecastListener, city: String) {

        val call = service.getForecastWeather(city)
        call.enqueue(object : Callback<ForecastRetrofit> {
            override fun onResponse(call: Call<ForecastRetrofit>, response: Response<ForecastRetrofit>) {
                if (response.body() != null && response.isSuccessful) {
                    listener.onSuccessResponse(response)
                } else {
                    listener.onFailureResponse(response.message())
                }
            }

            override fun onFailure(call: Call<ForecastRetrofit>, t: Throwable) {}
        })
    }
}